File: KiraVM/vm.cpp
#include <cstdio>
#include <cstdint>
#include <cstdlib>

#include "error.h"

#include "stack.h"
#include "internal.h"

#include "vm.h"

static Literal * locals;
static int32_t localCount;

String ** strings;
static int32_t stringCount;

static int64_t programLength;
static int32_t programOffset;

static int32_t readOffset = 4;

static int32_t getInt32(const unsigned char * buffer)
{
    auto ret = *((int32_t *) (buffer + readOffset));
    readOffset += 4;
    return ret;
}

static int64_t getInt64(const unsigned char * buffer)
{
    auto ret = *((int64_t *) (buffer + readOffset));
    readOffset += 8;
    return ret;
}

static float getFloat(const unsigned char * buffer)
{
    auto ret = *((float *) (buffer + readOffset));
    readOffset += 4;
    return ret;
}

static Literal getLiteral(const unsigned char * buffer)
{
    Literal ret{};
    auto type = static_cast<LiteralType>(*((int8_t *) (buffer + readOffset)));
    readOffset += 1;

    ret.type = type;

    switch(type)
    {
        case LiteralType::Integer:
            ret.i = getInt32(buffer);
            break;
        case LiteralType::Float:
            ret.f = getFloat(buffer);
            break;
        case LiteralType::String:
            ret.s = getInt32(buffer);
            break;
    }

    return ret;
}

static void setLiteral(int32_t index, Literal value)
{
    if(index < 0 || index > localCount)
        throwError("Invalid local index.");

    locals[index] = value;
}

void vm_initialize(const unsigned char * buffer)
{
    localCount = getInt32(buffer);
    locals = static_cast<Literal *>(calloc(static_cast<size_t>(localCount), sizeof(Literal)));

    stringCount = getInt32(buffer);
    strings = static_cast<String **>(malloc(sizeof(String) * stringCount));

    for(int i = 0; i < stringCount; i++)
    {
        auto byteLength = getInt32(buffer);
        auto str = utf8_create(buffer + readOffset, byteLength);
        strings[i] = str;
        readOffset += byteLength;
    }

    programLength = getInt64(buffer);
    programOffset = readOffset;

    stack_initialize();
}

void vm_run(const unsigned char * buffer)
{
    while(readOffset - programOffset < programLength)
    {
        auto instruction = static_cast<Instruction>(getInt32(buffer));

        switch(instruction)
        {
            case Instruction::Push:
            {
                auto value = getLiteral(buffer);

                stack_push(value);
            }
                break;
            case Instruction::Pop:
                stack_pop();
                break;
            case Instruction::Set:
            {
                auto index = getInt32(buffer);
                auto value = getLiteral(buffer);

                setLiteral(index, value);
            }
                break;
            case Instruction::CallS:
                break; // TODO: Implement
            case Instruction::CallI:
            {
                auto internal = getInt32(buffer);

                internal_execute(internal);
            }
                break;
            case Instruction::CallE:
                break; // TODO: Implement
        }
    }
}

void vm_deinitialize()
{
    stack_deinitialize();

    free(strings);
    free(locals);
}

File: KiraVM/src/vm.rs
use std::io::{Cursor, Read};

use internal;
use literal::Literal;
use read::{read_i32, read_i64, read_string, read_literal};

const INSTRUCTION_PUSH: i32 = 0x01;
const INSTRUCTION_POP: i32 = 0x02;

const INSTRUCTION_SET: i32 = 0x03;

const INSTRUCTION_CALLS: i32 = 0x10;
const INSTRUCTION_CALLI: i32 = 0x11;
const INSTRUCTION_CALLE: i32 = 0x12;

pub fn run(mut cursor: Cursor<Vec<u8>>) {
	let local_count = read_i32(&mut cursor) as usize;
	let mut locals: Vec<Literal> = Vec::with_capacity(local_count);

	for _ in 0..local_count {
		locals.push(Literal::Integer(0));
	}

	let string_count = read_i32(&mut cursor) as usize;
	let mut strings: Vec<String> = Vec::with_capacity(string_count);

	for _ in 0..string_count {
		strings.push(read_string(&mut cursor));
	}

	let program_length = read_i64(&mut cursor);
	let mut program = vec![0; program_length as usize];
	let mut stack: Vec<Literal> = Vec::new();

	cursor.read_exact(&mut program).expect("Failed to read program");

	let mut cursor = Cursor::new(program);

	while cursor.position() < program_length as u64 {
		let instruction = read_i32(&mut cursor);

		match instruction {
			INSTRUCTION_PUSH => {
				stack.push(read_literal(&mut cursor, &strings));
			},
			INSTRUCTION_POP => {
				stack.pop();
			},

			INSTRUCTION_SET => {
				let index = read_i32(&mut cursor) as usize;
				let value = read_literal(&mut cursor, &strings);

				locals[index] = value;
			},

			INSTRUCTION_CALLS => {} // TODO: Implement
			INSTRUCTION_CALLI => {
				internal::execute(read_i32(&mut cursor), &mut stack);
			},
			INSTRUCTION_CALLE => {} // TODO: Implement

			_ => { throw_error!("Got unknown instruction: {}", instruction); }
		};
	}
}
